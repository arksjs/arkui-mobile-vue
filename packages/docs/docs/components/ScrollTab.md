# ScrollTab/ScrollTabItem 标签滚动布局

<CodeDemo name="ScrollTab">

<<< @/../../demo/src/components/Navigation/ScrollTab/index.vue

</CodeDemo>

## Import

```js
import { AkScrollTab, AkScrollTabItem } from 'arkui-mobile-vue'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { ScrollTabOnChange } from 'arkui-mobile-vue'
```

## ScrollTab Props

| 属性                 | 类型             | 默认值 | 必填 | 说明                        |
| -------------------- | ---------------- | ------ | ---- | --------------------------- |
| sticky-offset-top    | string \| number | 0      | 否   | 数值默认是 px，也支持 vw/vh |
| sticky-offset-bottom | string \| number | 0      | 否   | 数值默认是 px，也支持 vw/vh |

## ScrollTab Events

| 事件   | 描述       | 回调函数参数                        | TypeScript 函数   |
| ------ | ---------- | ----------------------------------- | ----------------- |
| change | 切换时触发 | (name: string, activeIndex: number) | ScrollTabOnChange |

## ScrollTab Slots

### #default

注：其中只可放置 [ScrollTabItem](./ScrollTab.md#scrolltabitem-props) 组件，否则会导致未定义的行为。

```vue
<ak-scroll-tab>
  <ak-scroll-tab-item name="Dust Red">
    <div class="scroll-tab-box box-1"></div>
  </ak-scroll-tab-item>
  <ak-scroll-tab-item name="Volcano">
    <div class="scroll-tab-box box-2"></div>
  </ak-scroll-tab-item>
  <ak-scroll-tab-item name="Sunset Orange">
    <div class="scroll-tab-box box-3"></div>
  </ak-scroll-tab-item>
  ...
</ak-scroll-tab>
```

## ScrollTabItem Props

| 属性  | 类型   | 默认值 | 必填 | 说明                                                               |
| ----- | ------ | ------ | ---- | ------------------------------------------------------------------ |
| name  | string |        | 是   | 唯一标识，设置后配合 ScrollTab 组件的 `v-model` 和 `onChange` 使用 |
| title | string |        | 否   | 分组名，也应用于吸附，如果没有设置则获取 `name` 的值               |

## ScrollTabItem Slots

### #default

```vue
<ak-scroll-tab-item name="Dust Red">
  <div class="scroll-tab-box box-1"></div>
</ak-scroll-tab-item>
```
