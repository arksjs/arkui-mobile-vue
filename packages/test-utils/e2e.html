<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
