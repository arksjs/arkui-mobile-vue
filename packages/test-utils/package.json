{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
