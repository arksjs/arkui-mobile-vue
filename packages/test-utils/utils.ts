export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
