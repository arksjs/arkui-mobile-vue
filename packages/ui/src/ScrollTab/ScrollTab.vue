<template>
  <div :class="classes">
    <div class="ta-scroll-tab_bar">
      <Sticky
        ref="sideRef"
        :offsetTop="stickyOffsetTop"
        :offsetBottom="sideBar ? stickyOffsetBottom : undefined"
      >
        <Tab
          v-if="!sideBar && tabList.length > 0"
          :options="tabList"
          :modelValue="activeName"
          @change="onTabChange"
        />
        <SideTab
          v-else-if="sideBar && tabList.length > 0"
          :options="tabList"
          :modelValue="activeName"
          @change="onTabChange"
        />
      </Sticky>
    </div>
    <div class="ta-scroll-tab_body">
      <StickyView
        :offsetTop="viewOffsetTop"
        :modelValue="modelValue"
        :disabledHeader="!sideBar"
        :enablePullRefreshUp="enablePullRefreshUp"
        :enablePullRefreshDown="enablePullRefreshDown"
        :pullRefreshTexts="pullRefreshTexts"
        ref="bodyRef"
        @resetItems="onResetItems"
        @change="onStickyViewChange"
        @pullRefreshing="onPullRefreshing"
      >
        <slot></slot>
      </StickyView>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref, shallowRef, watch, computed, type PropType } from 'vue'
import { Tab } from '../Tab'
import { SideTab } from '../SideTab'
import { Sticky } from '../Sticky'
import { StickyView } from '../StickyView'
import { isSizeValue, isString, getSizeValue, type PropsToEmits } from '../helpers'
import type { StickyViewOnResetItems, StickyViewRef, StickyViewOnChange } from '../StickyView/types'
import { emitChangeValidator } from '../StickyView/props'
import type { ResetContainer, StickyRef } from '../Sticky/types'
import type {
  ScrollTabEmits,
  ScrollTabOnPullRefreshing,
  ScrollTabPullRefreshTexts,
  ScrollTabRef
} from './types'
import type { SideTabOnChange } from '../SideTab/types'
import { getClasses } from './util'
import { TAB_HEIGHT } from '../Tab/util'
import { emitRefreshingValidator } from '../ScrollView/props'

export default defineComponent({
  name: 'ta-scroll-tab',
  components: { Tab, SideTab, Sticky, StickyView },
  props: {
    modelValue: {
      type: String
    },
    stickyOffsetTop: {
      type: [Number, String],
      validator: isSizeValue,
      default: 0
    },
    stickyOffsetBottom: {
      type: [Number, String],
      validator: isSizeValue,
      default: 0
    },
    sideBar: {
      type: Boolean,
      default: true
    },
    // 允许上拉刷新
    enablePullRefreshUp: {
      type: Boolean,
      default: false
    },
    // 允许下拉刷新
    enablePullRefreshDown: {
      type: Boolean,
      default: false
    },
    // 下拉刷新提示文案
    pullRefreshTexts: {
      type: Object as PropType<ScrollTabPullRefreshTexts>
    },
    // 滚动挂在到document上
    documentContainer: {
      type: Boolean,
      default: true
    }
  },
  emits: {
    'update:modelValue': name => isString(name),
    change: emitChangeValidator,
    pullRefreshing: emitRefreshingValidator
  } as PropsToEmits<ScrollTabEmits>,
  setup(props, { emit, expose }) {
    const sideRef = shallowRef<StickyRef | null>(null)
    const bodyRef = shallowRef<StickyViewRef | null>(null)
    const tabList = ref<
      {
        value: string
        label: string
      }[]
    >([])
    const activeName = ref<string>()

    // 单独更新以下tab的activeName
    function updateActiveName(name?: string) {
      if (name != null && nameInList(name) && name !== activeName.value) {
        activeName.value = name
      }
    }

    function nameInList(name: string) {
      for (let i = 0; i < tabList.value.length; i++) {
        if (tabList.value[i].value === name) {
          return true
        }
      }

      return false
    }

    const onTabChange: SideTabOnChange = (name, index) => {
      scrollToIndex(index)
    }

    const onStickyViewChange: StickyViewOnChange = (name, index) => {
      updateActiveName(name)
      emit('update:modelValue', name)
      emit('change', name, index)
    }

    /**
     * 滚动到第index个
     */
    function scrollToIndex(index: number) {
      bodyRef.value?.scrollToIndex(index)
    }

    /**
     * 滚动到指定name
     */
    function scrollTo(name: string) {
      bodyRef.value?.scrollTo(name)
    }

    const resetContainer: ResetContainer = containSelector => {
      sideRef.value?.resetContainer(containSelector)
      bodyRef.value?.resetContainer(containSelector)
    }

    const onResetItems: StickyViewOnResetItems = items => {
      tabList.value = items.map(item => {
        return {
          value: item.name,
          label: item.title || item.name
        }
      })
    }

    const viewOffsetTop = computed(
      () => getSizeValue(props.stickyOffsetTop) + (props.sideBar ? 0 : TAB_HEIGHT)
    )

    const onPullRefreshing: ScrollTabOnPullRefreshing = (payload, loadComplete) => {
      emit('pullRefreshing', payload, loadComplete)
    }

    const classes = computed(() => getClasses(props.sideBar))

    watch(
      () => props.modelValue,
      val => updateActiveName(val)
    )

    onMounted(() => {
      resetContainer(props.documentContainer ? document.documentElement : undefined)
      updateActiveName(props.modelValue)
      if (activeName.value == null && tabList.value.length > 0) {
        // 首次要写入一个
        activeName.value = tabList.value[0].value
      }
    })

    expose({
      scrollTo,
      scrollToIndex,
      resetContainer
    } as ScrollTabRef)

    return {
      viewOffsetTop,
      classes,
      sideRef,
      bodyRef,
      activeName,
      tabList,
      onTabChange,
      onStickyViewChange,
      onResetItems,
      onPullRefreshing,

      scrollTo,
      scrollToIndex,
      resetContainer
    }
  }
})
</script>
