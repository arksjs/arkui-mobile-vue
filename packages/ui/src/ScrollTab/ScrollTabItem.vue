<template>
  <div class="fx-sticky-view-item fx-scroll-tab-item" :data-name="name">
    <div class="fx-sticky-view-item_header">{{ name }}</div>
    <div class="fx-sticky-view-item_body"><slot></slot></div>
  </div>
</template>

<script lang="ts">
import { defineComponent, inject, onMounted, onUnmounted } from 'vue'
import { createUpdateInItem } from '../hooks/use-list'

export default defineComponent({
  name: 'fx-scroll-tab-item',
  props: {
    name: {
      type: String,
      required: true
    }
  },
  setup() {
    const update = inject(
      'fxStickyViewUpdate',
      createUpdateInItem('scroll-tab')
    )

    onMounted(update)
    onUnmounted(update)

    return {}
  }
})
</script>
