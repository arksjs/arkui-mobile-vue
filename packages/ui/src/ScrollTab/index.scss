@import '../style/var.scss';

.ak-scroll-tab {
  display: flex;
  overflow-y: auto;
  position: relative;
  background-color: $background-color;

  &_sidebar {
    width: 80px;
    height: 100%;
    flex-grow: 0;
    background-color: $background-color;
  }

  &_body {
    flex: 1;
  }
}
