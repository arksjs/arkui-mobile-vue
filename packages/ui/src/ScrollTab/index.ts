import { multiWithInstall } from '../helpers/with-install'
import ScrollTab from './ScrollTab.vue'
import ScrollTabItem from './ScrollTabItem.vue'

export { ScrollTab, ScrollTabItem }
export default ScrollTab
export const install = multiWithInstall(ScrollTab, [ScrollTabItem])
