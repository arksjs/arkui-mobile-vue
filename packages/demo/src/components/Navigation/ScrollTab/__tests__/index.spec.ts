import { mount } from '@vue/test-utils'
import { ScrollTab, ScrollTabItem } from '@/ScrollTab'

describe('ScrollTab', () => {
  test('ScrollTab should render default correctly', () => {
    const wrapper = mount(ScrollTab, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('ScrollTabItem should render default correctly', () => {
    const wrapper = mount(ScrollTabItem, {
      props: { name: 'item-1' },
      slots: {
        default: 'content'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
