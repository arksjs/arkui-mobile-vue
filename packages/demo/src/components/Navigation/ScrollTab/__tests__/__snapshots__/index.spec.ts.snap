// Vitest Snapshot v1

exports[`ScrollTab > ScrollTab should render default correctly 1`] = `
"<div class=\\"fx-scroll-tab\\">
  <div class=\\"fx-scroll-tab_sidebar\\">
    <div class=\\"fx-sticky\\">
      <div class=\\"fx-sticky_content\\">
        <div class=\\"fx-side-tab\\">
          <ul class=\\"fx-side-tab_list\\"></ul>
        </div>
      </div>
    </div>
  </div>
  <div class=\\"fx-scroll-tab_body\\">
    <div class=\\"fx-sticky-view\\">
      <div class=\\"fx-sticky-view_list\\"></div>
      <div class=\\"fx-sticky fx-sticky-view_top\\">
        <div class=\\"fx-sticky_content\\">
          <div class=\\"fx-sticky-view_fixed\\">
            <div class=\\"fx-sticky-view_fixed-inner\\" style=\\"transform: translate3d(0px, -100%, 0px);\\"></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`ScrollTab > ScrollTabItem should render default correctly 1`] = `
"<div class=\\"fx-sticky-view-item fx-scroll-tab-item\\" data-name=\\"item-1\\">
  <div class=\\"fx-sticky-view-item_header\\">item-1</div>
  <div class=\\"fx-sticky-view-item_body\\">content</div>
</div>"
`;
